opentelemetry-sdk/src/opentelemetry/sdk/metrics/view.py

```diff
@@ -0,0 +1,107 @@
+
+
+class View:
+
+    def __init__(self, metric, aggregator_type, label_keys=None):
+        self.metric = metric
+        self.aggregator_type = aggregator_type
+        self._view_datas = {} # Labels to ViewData
+        self._view_datas_lock = threading.Lock()
+        # TODO: add configuration for aggregators (histogram, etc.)
+        # TODO: add label key configuration
+        self.label_keys = label_keys
+        if self.label_keys is None:
+            self.label_keys = []
+
+    def get_all_view_data(self):
+        # We return a copy of the values because the dictionary can be can be
+        # always changing
+        return list(self._view_datas.values())
+
+    def get_view_data(self, labels):
+        return self._view_datas.get(labels)
+
+    def update_view_datas(self, labels, value):
+        with self._view_datas_lock:
+            view_data = self.get_view_data(labels)
+            if view_data is None:
+                view_data = ViewData(labels, self.aggregator_type)
+                self._view_datas[labels] = view_data
+            view_data.aggregator.update(value)
+
+    # Uniqueness of View is based on metric and aggregation type
+    def __hash__(self):
+        return hash((self.metric, self.aggregator_type))
+
+    def __eq__(self, other):
+        return self.metric == other.metric and \
+            self.aggregator_type == other.aggregator_type
+
+
+class _ViewManager:
+
+    def __init__(self):
+        self._views = {} # metric to set of views
+        self._view_lock = threading.Lock()
+
+    def register_view(self, view):
+        with self._view_lock:
+            if self._views.get(view.metric) is None:
+                self._views[view.metric] = {view}
+            elif view not in self._views.get(view.metric):
+                self._views[view.metric].add(view)
+            else:
+                logger.warning("View already registered.")
+
+    def get_views(self, metric):
+        return self._views.get(metric, {})
+
+    def get_all_view_data_for_metric(self, metric):
+        all_view_data = []
+        for view in self.get_views(metric):
+            all_view_data.extend(view.get_all_view_data())
+        return all_view_data
+
+    def unregister_view(self, view):
+        with self._view_lock:
+            if self._views.get(view.metric) is None:
+                logger.warning("Metric for view does not exist.")
+            elif view in self._views.get(view.metric):
+                self._views.get(view.metric).remove(view)
+
+    def update_view(self, metric, labels, value):
+        with self._view_lock:
+            views = self.get_views(metric)
+            # We only update if view was defined for the metric
+            for view in views:
+                # TODO: Check view configuration here to keep/drop keys
+                # Find the view data corresponding to the labels
+                view.update_view_datas(labels, value)
+
+
+view_manager = _ViewManager()
```

Comment:
You might consider scoping the view manager to `Meter` or `MetricProvider` instead.

The use case for the view manager is this: given a `Meter`, get `{(metric, labels): aggregated_value}` for all metrics created via that `Meter` and checkpoint all those metrics' registered aggregations.

Right now this [happens in `Meter.collect`](https://github.com/lzchen/opentelemetry-python/blob/aa7562a556884e1402edbaaac97949bb7a8a7ed8/opentelemetry-sdk/src/opentelemetry/sdk/metrics/__init__.py#L276-L284). As an aside, as it's written now I think this doesn't guarantee that all metrics are actually collected at the same time.

If the view manager belonged to the meter instead you could return the full set of aggregators (or aggregation data as above) on each call instead of filtering by metric.

Reply:
I agree it makes more sense for the ViewManager to belong to the meter, this way we can iterate through all the metrics (since it is guaranteed to be the metrics that belong to that meter). I will make this change. One question I have is concerning your point: "as it's written now I think this doesn't guarantee that all metrics are actually collected at the same time". Can you explain why this is? The view manager is essentially just finding the view datas by metric, and these metrics belong to the meter that called collect in the first place.

Reply:
Just because there's no lock (or clever lock-free alternative) around `_collect_metrics`. The user can record new values on the main thread while the meter is iterating through views and metrics on the controller thread.

I don't expect this to be a big deal in practice, but it does have some potentially surprising consequences. In the main thread we might do:

```
counter_a.add(1)
counter_b.add(1)
```

And in the controller thread we might pick up the update to `counter_b` but not `counter_a` depending on the view or metric iteration order.

In the future we might also want a way for the user to say "include both or neither of these updates in this collection interval: do not split them into two different intervals", but that's a problem for another PR.